gui/src/main/java/kcl/teamIndexZero/traffic/gui/GUI_Primitives.java

```diff
@@ -0,0 +1,85 @@
+package kcl.teamIndexZero.traffic.gui;
+
+import kcl.teamIndexZero.traffic.log.Logger_Interface;
+
+import javax.imageio.ImageIO;
+import java.awt.*;
+import java.awt.geom.AffineTransform;
+import java.awt.geom.QuadCurve2D;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Created by kumar awijeet on 2/24/2016. thanks for Working!!
+ */
+public class GUI_Primitives {
+    protected static Logger_Interface log;
+    InputStream imageStream;
+    int j = 0, k = 0;
+    Image image, image1;
+    // TODO add log instance here.
+    public void drawLine(int x1, int y1, int x2, int y2, Graphics g) {
+        Graphics2D g2 = (Graphics2D) g;
+        g.drawLine(x1, y1, x2, y2);
+    }
+
+    public void drawCurve(int x1, int y1, int x2, int y2, int ctrlx, int ctrly, Graphics g) {
+        Graphics2D g2 = (Graphics2D) g;
+        QuadCurve2D q = new QuadCurve2D.Float();
+        q.setCurve(x1, y1, ctrlx, ctrly, x2, y2);
+        g2.draw(q);
+    }
+
+    public void drawSingleRoad(int x, int y, int x1, int y1, int x2, int y2, int x3, int y3, Graphics g) {
+        drawLine(x, y, x1, y1, g);
+        drawLine(x2, y2, x3, y3, g);
+    }
+
+    public void drawSmallCar(int x3, int y3, double angleToXAxis, Graphics g) {
+        drawCar(x3, y3, angleToXAxis, "gui/src/main/resources/sprites/bmw_z_top_view_clip_art_18132.jpg", g);
+    }
+
+    public void drawTruck(int x3, int y3, double angleToXAxis, Graphics g) {
+        drawCar(x3, y3, angleToXAxis, "gui/src/main/resources/sprites/8460_st0640_117.jpg", g);
+    }
+
+    private void drawCar(int x3, int y3, double angleToXAxis, String filename, Graphics g) {
+        /*
+        Files taken from
+        http://all-free-download.com/free-vector/car-vector-top-view-download.html
+        http://hdimagelib.com/trailer+truck+top+view
+        */
+        Graphics2D graphics = (Graphics2D) g;
+        AffineTransform originalTransform = graphics.getTransform();
+        try {
+            AffineTransform newTransformation = new AffineTransform();
+            newTransformation.rotate(angleToXAxis, x3, y3);
+            graphics.setTransform(newTransformation);
+            //InputStream imageStream;
+            if ((filename == "gui/src/main/resources/sprites/bmw_z_top_view_clip_art_18132.jpg") && (k == 0)) {
```

Comment:
Alex tried your image == null check, and somehow it takes the time back up, so it's better to leave it with the flags. I think this system works properly with respect to time, let it be that way, let's see other components, like I wanted to check if the flags were actually working, so will create a test for them and check that. Plus we also have to add logging for other exceptions. Also finish the finally branch. So let us see those sections. ‎Moreover, it's not like this code isn't working, so my suggestion is let's wrap it up and finish the rest so that we are able to get it merged and give time to report and other miscellaneous things as well.                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone on the EE network.                                                                                                                                                                                                                From: Alexander CherednichenkoSent: Friday 25 March 2016 19:59To: teamIndexZero/index_zero_trafficsystemReply To: teamIndexZero/index_zero_trafficsystemCc: kumar awijeetSubject: Re: [index_zero_trafficsystem] Gui graphics kumar awijeet (#9)In gui/src/main/java/kcl/teamIndexZero/traffic/gui/GUI_Primitives.java:
> +    }
> +
> +    private void drawCar(int x3, int y3, double angleToXAxis, String filename, Graphics g) {
> +        /*
> +        Files taken from
> +        http://all-free-download.com/free-vector/car-vector-top-view-download.html
> +        http://hdimagelib.com/trailer+truck+top+view
> +        */
> +        Graphics2D graphics = (Graphics2D) g;
> +        AffineTransform originalTransform = graphics.getTransform();
> +        try {
> +            AffineTransform newTransformation = new AffineTransform();
> +            newTransformation.rotate(angleToXAxis, x3, y3);
> +            graphics.setTransform(newTransformation);
> +            //InputStream imageStream;
> +            if ((filename == "gui/src/main/resources/sprites/bmw_z_top_view_clip_art_18132.jpg") && (k == 0)) {

You got the point! 
It should be faster now. 

Instead of using extra variable for the flag, you could do much easier thing: check if the image itself is null or not. I.e.: 

public class Primitives {
   ...
   private Image image1 = null;
   ...
   public drawCar(..., filename ) {
      if(image1 == null && filename = "/...car.png" ) {
          ... load image and file
          image1 = ImageIO.read(...)
      } 
      graphics.draw(... image);
}


However, you will need to maintain separate variable for each of the image, and check if filename == one or another. What you could do instead, is use Map<String, Image>, such as HashMap, to use store corresponding pairs of Path --> Image object, and the code would be like:

public class Primitives {
   private Map<String, BufferedImage> imageCache = new HashMap<>();
   ...
   public void drawCar(..., filename) {
       if(!imageCache.containsKey(filename)) {
         // load image, 
         BufferedImage im = ImageIO.read(...)
         imageCache.put(filename, im);
   }
   im = imageCache.get(filename) // we are sure now that it exists, as if it would have not, we would have already loaded it
   graphics.drawImage(im);
}


Does that make sense? 

—You are receiving this because you were mentioned.Reply to this email directly or view it on GitHub